dipy/reconst/ivim.py

```diff
@@ -509,6 +536,362 @@ def _leastsq(self, data, x0):
                 return x0
 
 
+class IvimModelVP(ReconstModel):
+
+    def __init__(self, gtab, fit_method='VarPro', *args, **kwargs):
+        r""" Initialize an IvimModelVP class.
+
+        The IVIM model assumes that biological tissue includes a volume
+        fraction 'f' of water flowing with a pseudo-diffusion coefficient
+        D* and a fraction (1-f) of static (diffusion only), intra and
+        extracellular water, with a diffusion coefficient D. In this model
+        the echo attenuation of a signal in a single voxel can be written as
+
+            .. math::
+
+            S(b) = S_0*[f*e^{(-b*D\*)} + (1-f)e^{(-b*D)}]
+
+            Where:
+            .. math::
+
+            S_0, f, D\* and D are the IVIM parameters.
+
+        Parameters
+        ----------
+        gtab : GradientTable class instance
+            Gradient directions and bvalues
+
+        fit_method : str or callable
+            str can be one of the following:
+
+            'LM' For Levenber-Marquardt fitting with 2-stage option
+                This method does the same as :func:`ivim.IvimModel(gtab,
+                                                                   fit_method=
+                                                                   'LM')`
+            'VP' For the Variable Projections based multi-stage fitting option
+                This method does the same as :func:`ivim.IvimModel(gtab,
+                                                                   fit_method=
+                                                                   'VP')`
+
+        tol : float, optional
+            Tolerance for convergence of minimization.
+            default : 1e-15
+
+        maxiter: int, optional
+            Maximum number of iterations for the Differential Evolution in
+            SciPy.
+            default : 10
+
+        References
+        ----------
+        .. [1] Le Bihan, Denis, et al. "Separation of diffusion and perfusion
+               in intravoxel incoherent motion MR imaging." Radiology 168.2
+               (1988): 497-505.
+        .. [2] Federau, Christian, et al. "Quantitative measurement of brain
+               perfusion with intravoxel incoherent motion MR imaging."
+               Radiology 265.3 (2012): 874-881.
+        .. [3] Fadnavis, Shreyas et.al. "MicroLearn: Framework for machine
+               learning, reconstruction, optimization and microstructure
+               modeling, Proceedings of: International Society of Magnetic
+               Resonance in Medicine (ISMRM), Montreal, Canada, 2019.
+        """
+
+        self.maxiter = kwargs.get('maxiter', 10)
+        self.xtol = kwargs.get('xtol', 1e-8)
+        self.bvals = gtab.bvals
+        self.yhat_perfusion = np.zeros(self.bvals.shape[0])
+        self.yhat_diffusion = np.zeros(self.bvals.shape[0])
+        self.exp_phi1 = np.zeros((self.bvals.shape[0], 2))
+
+    @multi_voxel_fit
+    def fit(self, data, mask=None):
```

Comment:
mask not needed the decorator can deal with that.

Reply:
Cool!